ideas/apq.h

```diff
@@ -0,0 +1,285 @@
+using io_service = boost::asio::io_service;
+
+
+/**
+* Should we use defined type of pg_connect_handle, or some customized type and
+* template for connection?
+*   PGconn                       - pointer to native connection handle
+*   std::function<void(PGconn*)> - finalization function, e.g. destructor for single 
+*                                  connection or recycling operation for pool.
+*/
+using pg_connect_handle = std::shared_ptr<PGconn>; //, std::function<void(PGconn*)>>;
+
+
+/**
+* We can try to use standard error_code, but we will loose the error message
+* from database. But if we will use the custom error_code we are forced to
+* implement our own yield_context and async_result infrastructure. This perspective
+* is not attractive. We can provide some kind of yield_context support, but with
+* message loss in case of error_code bound to yield (i.e. yield[ec])
+*/
+using error_code = boost::system::error_code; // std:: / custom ?
```

Comment:
Until we use ```boost::asio``` let's use ```boost::system_error```. For networking ts we'll back to this question.

Reply:
>  when we use standard error_code we loose an error message form PostgreSQL

why? can we not communicate that message through a custom error_category?

Reply:
> why? can we not communicate that message through a custom error_category?

Nope - error_category can not provide an ability for context depended information such as error messages from database which contains information about syntax error in query's text for example. So you can say "oh, you have a syntax error" with standard error_code, but you can not say "oh, you have a syntax error at column 45". "at column 45" is a context depended information, and there is no rooms for it in std or boos error_code. What's why we are using custom error_code in the meta access.